verilog/tools/extractor/verilog_extractor.cc

```diff
@@ -0,0 +1,332 @@
+    auto resolver = tagExtractor.find(node.Tag().tag);
+    if (resolver != tagExtractor.end()) {
+        blocks.push_back(resolver->second(node, node, base));
+        return blocks;
+    }
+
+    for (const auto &child : node.children()) {
+        if (child) {
+            if (child->Kind() == verible::SymbolKind::kNode) {
+                auto innerBlock = Extract(verible::SymbolCastToNode(*child), base);
+                blocks.insert(blocks.end(), innerBlock.begin(), innerBlock.end());
+            } else {
+                auto innerBlock = Extract(verible::SymbolCastToLeaf(*child), base);
+                blocks.insert(blocks.end(), innerBlock.begin(), innerBlock.end());
+            }
+        }
+    }
+
+    std::cout << "End Node" << std::endl << std::endl;
+    return blocks;
+}
+
+const verible::Symbol *GetChildByTag(const verible::SyntaxTreeNode &root, verilog::NodeEnum tag) {
+    for (const auto &child : root.children()) {
+        if (child) {
+            if (child->Tag().tag == static_cast<int>(tag)) {
+                return child.get();
+            }
+        }
+    }
+    return nullptr;
+}
+
+const verible::Symbol *GetFirstChildByTag(const verible::SyntaxTreeNode &root, verilog::NodeEnum tag) {
+    auto target = GetChildByTag(root, tag);
+    if (target == nullptr) {
+        for (const auto &child : root.children()) {
+            if (child && child->Kind() == verible::SymbolKind::kNode) {
+                target = GetFirstChildByTag(verible::SymbolCastToNode(*child), tag);
+                if (target != nullptr) {
+                    return target;
+                }
+            }
+        }
+    }
+    return target;
+}
+
+Block ExtractModuleInstantiation(const verible::SyntaxTreeNode &node, const verible::SyntaxTreeNode &root,
+                                 absl::string_view base) {
+
+    auto instantiationBase = GetChildByTag(node, verilog::NodeEnum::kInstantiationBase);
+
+    auto instantiationType = GetChildByTag(verible::SymbolCastToNode(*instantiationBase),
+                                           verilog::NodeEnum::kInstantiationType);
+
+    auto type = verible::SymbolCastToLeaf(
+            *GetFirstChildByTag(verible::SymbolCastToNode(*instantiationType),
+                                verilog::NodeEnum::kNetVariableDeclarationAssign));
+
+    Anchor typeAnchor = Anchor(type, base);
+
+    auto variableList = GetChildByTag(verible::SymbolCastToNode(*instantiationBase),
+                                      verilog::NodeEnum::kGateInstanceRegisterVariableList);
+    auto variableName = verible::SymbolCastToLeaf(
+            *GetFirstChildByTag(verible::SymbolCastToNode(*variableList),
+                                verilog::NodeEnum::kNetVariableDeclarationAssign));
+    Anchor variableNameAnchor = Anchor(variableName, base);
+
+    Block moduleInstance = Block(Type::ModuleInstance);
+    moduleInstance.AppendAnchor(typeAnchor);
+    moduleInstance.AppendAnchor(variableNameAnchor);
```

Comment:
A lot of the visitation boilerplate could be saved by using a `TreeContextVisitor`.

You could use that to build up a hierarchical tree of `Block`s in a stack-like manner that operates on `TheCurrentBlock()`.